text/primitive-projections.md

```diff
@@ -0,0 +1,99 @@
+
+This document is an attempt to propose solutions to clarify the picture.
+
+See also [#14640](https://github.com/coq/coq/pull/14640) which tries to set up a context to stabilize a design choice.
+
+## Issue 1: the relative status of `Const`-based projections and `Proj`-based projections
+
+It is common to write `f param t` to mean a compact `Proj` kernel node. It is also useful to have a non compact representation for partial application of a projection or for an unapplied projection. What to do with it?
+
+*Design choice 1*: The kernel provides only compact `Proj` nodes for projections in primitive record types. It is the responsibility of the user (or of the vernac level) to later declare `Definition f' params x := x.(f params)` (with a new name) and to use it as an ordinary definition.
+  - Advantages: a simple model
+  - Drawbacks: the name `f'` is different
+
+1. *Design subchoice 1.1*: `Proj` nodes are exclusively parsed and printed using the `x.(f)` without parameters.
+   - Advantages: a direct correspondence between concrete and internal syntax
+   - Drawbacks: possible confusion with the more general syntax for projections of non-primitive records which has the form `x.(f params)` and possible confusion with the "tolerance" for using at parsing time the notation `x.(f params)` for any constant `f` even if not a projection
+
+2. *Design subchoice 1.2*: Idem but using a different syntax such as `x.[t]`, or `x.((t))`, or `x.{t}`, etc.
+   - Advantages: no confusion with the existing uses of the `x.(f)` syntax
+   - Drawbacks: reserves a new syntactic construction
+
+3. *Design subchoice 1.3*: `Proj` nodes are exclusively parsed and printed using the `x.(f _ ... _)` syntax where the `_` take the place of the parameters, noticing however that if the parameters are declared implicit, this reduces to the notation `x.(f)`.
+   - Advantages: compatibility with the existing syntax for projections in non-primitive records
+   - Drawbacks: one may still want a clearer distinction from the other uses of the syntax `x.(f params)`
+
+4. *Design subchoice 1.4*: `Proj` nodes are exclusively printed using the `x.(f _ ... _)` syntax (reducing to `x.(f)` with implicit parameters) but it is tolerated to use `f params x` at parsing as a compatibility notation for `Proj(f,x)` (when `f` is a projection in a primitive record type).
+   - Advantages: compatibility with the existing usage of often writing `f params x` for a primitive projection
+   - Drawbacks: asymmetry parsing/printing (but deprecating `f params x` could eventually clarify things)
+
+*Design choice 2*: Whenever a record is defined, a `Const f` node of same name `f` is made available at the same time and `Proj (f,x)` and `App (Const f, params, x)` are considered convertible by default without needing any form of explicit unfolding. Note that the infrastructure to make them convertible basically already exists in coercions, unification, ...
```

Comment:
It would be nice if the terms were syntactically equal so that hash-consing and the like would identify these terms.